Source/Internal/IGListPreprocessor.m

```diff
@@ -0,0 +1,98 @@
+//
+//  IGListPreprocessor.m
+//  IGListKit
+//
+//  Created by Adlai Holler on 2/20/17.
+//  Copyright © 2017 Instagram. All rights reserved.
+//
+
+#import "IGListPreprocessor.h"
+
+#import "IGListAssert.h"
+#import "IGListPreprocessingContext.h"
+#import "IGListPreprocessingDelegate.h"
+#import "IGListSectionController.h"
+#import "IGListSectionMap.h"
+
+@implementation IGListPreprocessor {
+	NSMapTable<IGListPreprocessingContext *, id<IGListPreprocessingDelegate>> *_contextToDelegateMap;
+	IGListSectionMap *_sectionMap;
+	dispatch_group_t _group;
+	dispatch_block_t _completion;
+	BOOL _ranCompletion;
+}
+
+- (instancetype)initWithSectionMap:(IGListSectionMap *)sectionMap {
+	if (self = [super init]) {
+		_contextToDelegateMap = [NSMapTable strongToStrongObjectsMapTable];
+		_sectionMap = [sectionMap copy];
+		_group = dispatch_group_create();
+	}
+	return self;
+}
+
+- (void)schedulePreprocessingWithContainerSize:(CGSize)containerSize completion:(nonnull dispatch_block_t)completionBlock {
+	IGAssertMainThread();
+	_completion = completionBlock;
+	
+	dispatch_queue_t queue = dispatch_queue_create("IGListPreprocessor Queue", DISPATCH_QUEUE_CONCURRENT);
+	
+	dispatch_group_t group = _group;
+	NSMapTable *contextToDelegateMap = _contextToDelegateMap;
+	[_sectionMap enumerateUsingBlock:^(id  _Nonnull object, IGListSectionController<IGListSectionType> * _Nonnull sectionController, NSInteger section, BOOL * _Nonnull stop) {
+		id<IGListPreprocessingDelegate> delegate = sectionController.preprocessingDelegate;
+		if (delegate == nil) {
+			return;
+		}
+		
+		// Create a context and associate it with this delegate.
+		IGListPreprocessingContext *context = [[IGListPreprocessingContext alloc] initWithObject:object containerSize:containerSize sectionIndex:section dispatchGroup:group];
+		[contextToDelegateMap setObject:delegate forKey:context];
+		
+		// Schedule the delegate's work – the context will call preprocessingDidCompleteWithContext:
+		dispatch_group_enter(group);
+		dispatch_async(queue, ^{
+			[delegate preprocessWithContext:context];
+		});
+	}];
+	
+	// When the work is all done, call -runCompletionIfNeeded on main.
+	dispatch_group_notify(_group, dispatch_get_main_queue(), ^{
+		[self runCompletionIfNeeded];
+	});
+}
+
+- (void)waitForPreprocessingToFinish {
```

Comment:
This seems a little dangerous, right? Allowing someone on the main thread to block until a bg queue completes can result in deadlocks/priority inversion.